pkg/metadata/service_discovery.go

```diff
@@ -15,12 +15,56 @@
 package metadata
 
 import (
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/common/model"
+
 	"github.com/parca-dev/parca-agent/pkg/discovery"
 )
 
 // ServiceDiscovery metadata provider.
-func ServiceDiscovery(m *discovery.Manager) *Provider {
-	return &Provider{
-		"service discovery", m.ProcessLabels,
+func ServiceDiscovery(logger log.Logger, m *discovery.Manager) Provider {
+	provider := &StatefulProvider{
+		name:  "service discovery",
+		state: map[int]model.LabelSet{},
+		mtx:   &sync.RWMutex{},
 	}
+
+	go func() {
+		defer level.Warn(logger).Log("msg", "service discovery metadata provider exited")
+
+		for tSets := range m.SyncCh() {
+			state := map[int]model.LabelSet{}
+			// Update process labels.
+			for _, groups := range tSets {
+				for _, group := range groups {
+					for _, pid := range group.PIDs {
+						// Overwrite the information we have here with the latest.
+						allLabels := model.LabelSet{}
+						for k, v := range group.Labels {
+							allLabels[k] = v
+						}
+						for _, t := range group.Targets {
+							for k, v := range t {
+								allLabels[k] = v
+							}
+						}
+
+						_, ok := state[pid]
+						if ok {
+							state[pid] = state[pid].Merge(allLabels)
```

Comment:
Good question. I don't have an answer off the top of my head. I assumed it wouldn't be the case; I'll double-check.

Reply:
I think I will keep this in place. This shouldn't have happened right now. `systemd` and `kubernetes` discovery mechanisms don't overlap. However, there could be cases in the future this could happen, so I'm leaning to keep this in place.